src/WebJobs.Script.WebHost/TokenExpiration/TokenExpirationService.cs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.PortableExecutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.Auth;
+using Microsoft.Azure.WebJobs.Script.Diagnostics;
+using Microsoft.Azure.WebJobs.Script.Properties;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.TokenExpiration
+{
+    internal class TokenExpirationService : IHostedService, IDisposable
+    {
+        private readonly IEnvironment _environment;
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly IOptionsMonitor<StandbyOptions> _standbyOptionsMonitor;
+        private readonly WebJobsScriptHostService _scriptHost;
+        private CancellationTokenSource _cancellationTokenSource;
+        private Task _analysisTask;
+        private bool _disposed;
+        private bool _analysisScheduled;
+        private ILogger _logger;
+
+        public TokenExpirationService(IEnvironment environment, WebJobsScriptHostService scriptHost, ILoggerFactory loggerFactory, IOptionsMonitor<StandbyOptions> standbyOptionsMonitor)
+        {
+            _environment = environment;
+            _scriptHost = scriptHost;
+            _loggerFactory = loggerFactory;
+            _standbyOptionsMonitor = standbyOptionsMonitor;
+            _logger = _loggerFactory.CreateLogger<TokenExpirationService>();
+        }
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (_environment.IsCoreTools())
+                {
+                    return Task.CompletedTask;
+                }
+                if (_standbyOptionsMonitor.CurrentValue.InStandbyMode)
+                {
+                    _standbyOptionsMonitor.OnChange(standbyOptions =>
+                    {
+                        if (!standbyOptions.InStandbyMode && !_analysisScheduled)
+                        {
+                            ScheduleTokenExpirationCheck();
+                        }
+                    });
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error starting token expiration service. Handling error and continuing.");
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                _cancellationTokenSource?.Cancel();
```

Comment:
@brettsam @jviau would it make sense to check if _disposed is true before trying to cancel? Not sure if there is any potential for a race here that could be addressed with something like `lock (this) if (!_disposed) _cancellationTokenSource?.Cancel();` and a similar lock in Dispose()

Reply:
I don't believe there is any thread-safety requirement to `IHostedService`. `.Dispose()` and `.StopAsync()` shouldn't be called together. It will be Stop -> Dispose.

With that said, there is a different issue: Dispose -> Stop will cause issues, as `CancellationTokenSource.Cancel` will throw. I typically always cancel my CTS's during dispose. But not a big concern here.

Reply:
Removed the StopAsync since `TokenExpirationService` now implements `BackgroundService`